### ISSUE TITLE
clarify Rotate trait: center vs. centroid

### ISSUE THREAD

**Issue by michaelkirk** [georust/geo#657]
Per some discussion in https://github.com/georust/geo/pull/656/files#r676937398, the expected behavior of Rotate isn't super clear.

Leaning on shapely, we came upon this more explicit and flexible api:

```
trait Rotate {
  // this is the existing method. Existing implementations vary, but most rotate around the geometries centroid.
  // let's deprecate it, and (I guess?) have it call centroid since that's what most impls currently do. 
  #[deprecate(note="equivalent to `rotate_around_centroid`, call that instead, or `rotate_around_center` if you'd like to rotate around the geometry's bounding box center")]
  fn rotate(&self, angle: T) -> Self {
    self.rotate_around_centroid(angle)
  }

  fn rotate_around_coord(&self, coord: Coordinate<T>, angle: T) -> Self;

  fn rotate_around_center(&self, angle: T) -> Self {
    let center = g.bounding_rect().center();
    g.rotate_around_coord(center, angle)
  }

  fn rotate_around_centroid(&self, angle: T) -> Self {
    let centroid = self.centroid();
    self.rotate_around_coord(centroid, angle)
  }
}
```

**Comment by michaelkirk** [comment]
@mapoulos, @urschrei  (and anyone else interested) - WDYT of this proposal?

@mapoulos - If you were interested in continuing to work on `Rotate` stuff, you could incorporate fixing #655 as part of this, or do #655 separately and follow up with this later (or if you run out of time / interest, LMK and we can hand it off).

**Comment by mapoulos** [comment]
I like the idea of deprecating `rotate` and adding `rotate_around_centroid`. 

My only hesitation is that `center` vs `centroid` is kind of confusing for a newbie. (Here's an example of an turf user confused by the numerous options:
 https://stackoverflow.com/questions/55982479/difference-between-centroid-and-centerofmass-in-turf ).  

It looks like `turfjs` and `shapely` are agreed about what a `center` is though, so it's probably safe to use [turfjs also uses a bbox for center](https://github.com/Turfjs/turf/blob/master/packages/turf-center/index.ts)

**Comment by mapoulos** [comment]
And I'm happy to do #655 and #657 together.

**Comment by michaelkirk** [comment]
> My only hesitation is that center vs centroid is kind of confusing for a newbie. (Here's an example of an turf user confused by the numerous options:

Oh interesting! Apparently our "centroid" is more like a turfjs's, "centerOfMass", meaning the average of "all points" in the shape - an area calculation, whereas turfjs's "centroid" is an average of the vertices.

I'll note that our definition of centroid agrees with every other source I can find:
- JTS (and thus [GEOS](http://geos.refractions.net/ro/doxygen_docs/html/classgeos_1_1algorithm_1_1CentroidArea.html), thus shapely)
- http://www.faqs.org/faqs/graphics/algorithms-faq/
- https://en.wikipedia.org/wiki/Centroid

It seems to me like turf.js is out on a ledge here, and I'd personally be comfortable breaking with them, and aligning with shapely/jts/geos.

What do others think?

**Comment by urschrei** [comment]
I…did not know that the definition of centroid was contested. I (think?) I wrote our current implementation and based it on Shapely / GEOS, and we should conform to those implementations IMO.

**Comment by mapoulos** [comment]
Yeah agreed that `turfjs` is on a limb here.

**Comment by mapoulos** [comment]
One further question:

Do we want to preserve the current behavior of polygon rotation in the `rotate_around_centroid` method? For true backwards compatibility we'd need to keep rotate so that it ignores internal holes on rotation, whereas I'd assume that `rotate_around_centroid` should always use the true centroid.

**Comment by mapoulos** [comment]
So the question is: is ignoring internal holes behavior worth preserving? it does diverge from what the docs say the method does. 

> /// Rotate a Geometry around its centroid by an angle, in degrees
    ///
    /// Positive angles are counter-clockwise, and negative angles are clockwise rotations.

**Comment by michaelkirk** [comment]
> Do we want to preserve the current behavior of polygon rotation in the rotate_around_centroid method? For true backwards compatibility we'd need to keep rotate so that it ignores internal holes on rotation, whereas I'd assume that rotate_around_centroid should always use the true centroid.
>
> So the question is: is ignoring internal holes behavior worth preserving? it does diverge from what the docs say the method does.

My opinion is that `rotate_around_centroid` should use the true centroid, accounting for the holes.

I'm in favor of deprecating the ambiguous `rotate` method, and deleting it in a future release. 

I think the trickier question is: until we delete the deprecated `rotate` method, what should it's behavior be? Do we maintain its current behavior, which ignores holes, or have it delegate to the new uniform centroid behavior. 

I don't have strong opinions there, but if forced to choose one, I'd guess that having it continue to ignore holes will surprise less people. So, if it's not difficult to maintain the legacy behavior for (multi)/polygons, I suppose I have a verrrrry slight preference to keep the existing behavior until the method is deleted, and maybe we can allude to this behavior in the methods deprecation notice.

Happy to be overruled on that decision if someone has a different inclination.

**Comment by mapoulos** [comment]
That’s my inclination too.

>

**Comment by urschrei** [comment]
I agree. We could even note what it does versus the correct implementation in its deprecation note.
